sdks/uniswapx-sdk/src/order/V3DutchOrder.ts

```diff
@@ -0,0 +1,514 @@
+import { SignatureLike } from "@ethersproject/bytes";
+import { PermitTransferFrom, PermitTransferFromData, SignatureTransfer, Witness } from "@uniswap/permit2-sdk";
+import { BigNumber, ethers } from "ethers";
+
+import { getPermit2, ResolvedUniswapXOrder } from "../utils";
+import { getBlockDecayedAmount } from "../utils/dutchBlockDecay";
+
+import { BlockOverrides, OffChainOrder, OrderInfo, V3DutchInput, V3DutchInputJSON, V3DutchOutput, V3DutchOutputJSON, V3OrderResolutionOptions } from "./types";
+
+import { originalIfZero } from ".";
+
+export type CosignerDataJSON = {
+    decayStartBlock: number;
+    exclusiveFiller: string;
+    exclusivityOverrideBps: number;
+    inputOverride: string;
+    outputOverrides: string[];
+}
+
+export type UnsignedV3DutchOrderInfoJSON = Omit<UnsignedV3DutchOrderInfo, "nonce" | "input" | "outputs" | "cosignerData"> & {
+    nonce: string;
+    input: V3DutchInputJSON;
+    outputs: V3DutchOutputJSON[];
+};
+
+export type CosignerData = {
+    decayStartBlock: number;
+    //No end in cosignerData
+    exclusiveFiller: string;
+    exclusivityOverrideBps: BigNumber;
+    inputOverride: BigNumber;
+    outputOverrides: BigNumber[];
+}
+
+export type UnsignedV3DutchOrderInfo = OrderInfo & {
+    cosigner: string;
+    input: V3DutchInput; //different from V2DutchOrder
+    outputs: V3DutchOutput[];
+};
+
+export type CosignedV3DutchOrderInfo = UnsignedV3DutchOrderInfo & {
+    cosignerData: CosignerData;
+    cosignature: string;
+};
+
+type V3WitnessInfo = {
+    info: OrderInfo,
+    cosigner: string,
+    baseInput: V3DutchInput,
+    baseOutputs: V3DutchOutput[],
+};
+
+
+const COSIGNER_DATA_TUPLE_ABI = "tuple(uint256,address,uint256,uint256,uint256[])";
+
+const V3_DUTCH_ORDER_TYPES = {
+    V3DutchOrder: [
+        { name: "info", type: "OrderInfo" },
+        { name: "cosigner", type: "address" },
+        { name: "baseInput", type: "V3DutchInput" },
+        { name: "baseOutputs", type: "V3DutchOutput[]" },
+    ],
+    OrderInfo: [
+        { name: "reactor", type: "address" },
+        { name: "swapper", type: "address" },
+        { name: "nonce", type: "uint256" },
+        { name: "deadline", type: "uint256" },
+        { name: "additionalValidationContract", type: "address" },
+        { name: "additionalValidationData", type: "bytes" },
+    ],
+    V3DutchInput: [
+        { name: "token", type: "address" },
+        { name: "startAmount", type: "uint256" },
+        { name: "curve", type: "V3Decay" },
```

Comment:
i know the contract abi is changing still but the type name must match what is in the contract https://github.com/Uniswap/UniswapX/pull/269/files#diff-ba488278e700ac9fd3b42d4d36a035810f7f76a75de5b9fc7eedad3136bd2ccbR53 , so "NonlinearDutchDecay"

Reply:
updated!